### DISCUSSION TITLE
想问一下 tg termes 字体设置中一部分代码的原理

### DISCUSSION THREAD

**Discussion by syvshc** [sjtug/SJTUThesis#802] [提问] [answered]
由于 elegantbook 近期在更换字体设置时遇到了问题: https://github.com/ElegantLaTeX/ElegantBook/issues/196 , 使用如下代码可以解决, 但是不是很清楚其原理, 想来询问一下, 感谢

https://github.com/sjtug/SJTUThesis/blob/b164d0f5b3087ed86c1038b5a85014b47a92c6f2/texmf/tex/latex/sjtuthesis/fd/sjtu-math-font-termes.def#L52-L65

**Comment by syvshc** [comment]
我的理解是将设置完 TG Termes 后的 `\encodingdefault, \rmdefault, \sfdefault, \ttdefault,` 保存下来, 即 `TU, texgyretermes, lmss, lmtt`, 然后下一步我不是很理解, 用  `fontenc` 宏包将编码转换为 `T1`, 然后我猜是设置了一些 newtx 系列字体, 因为我看到 `newtxtext` 宏包中有 `ntxlf, qhv, ntxtt` 这些字体的设置, 再加载 `newtxmath` 宏包, 然后恢复之前的 `\encodingdefault, \rmdefault, \sfdefault, \ttdefault,`. 或许我应该看一下 `newtxmath` 的源码……

**Comment by AlexaraWu** [comment, marked as answer]
这里的道理其实很简单，以下是 `newtxmath` 设置字体的部分代码：
```latex
\DeclareMathAlphabet{\mathsf}{\encodingdefault}{\sfdefault}{m}{n}
\DeclareMathAlphabet{\mathit}{\tx@enc}{\rmdefaultB}{m}{it}
\DeclareMathAlphabet{\mathtt}{\encodingdefault}{\ttdefault}{m}{n}
\DeclareSymbolFontAlphabet{\mathrm}{operators}
```
可以看到 `newtxmath` 会使用当前的 `\encodingdefault`, `\rmdefault`, `\sfdefault`, `\ttdefault` 来设置 `operators`, `\mathrm`, `\mathsf`, `\mathtt` 字体。

termes 字体设置这段代码做的就是在局部设置  `\encodingdefault`, `\rmdefault`, `\sfdefault`, `\ttdefault` 为 `newtxtext` 中的默认值，再加载  `newtxmath` 包。（`\RequirePackage[T1]{fontenc}` 的主要作用就是 `\def\encodingdefault{T1}`）

因为 `ctex` 包使用 `fontspec` 来设置 CJK 字体。为了不在正文环境中混用 OpenType 和 Type 1 字体，SJTUThesis 统一使用 `fontspec` 来设置正文中使用 OpenType 西文字体。因此我们不直接使用 `newtxtext` 包，只取其的默认值给 `newtxmath`。将数学环境中的 `operators` 等设置为风格一致的 Type 1 字体。

另外很重要的一点是 `fontspec` 在默认情况下，`\setmainfont` 会同时设置 `\mathrm`字体。而 SJTUThesis 的设计原则是正文字体和数学字体是分别独立设置的，互不干扰。所以要给 `fontspec` 添加 `no-math` 选项阻止这种行为。

**Reply by syvshc** [reply]
我明白了，顺便我发现 newtx 宏包在 v1.7 版本提供了一个新的 otf 字体，名称为 TeX Gyre TermesX，是否可以设置为正文字体来和 `newtxmath` 进行更好的一致性适配呢？但是貌似没有 sans 和 tt 版本的字体……

**Reply by AlexaraWu** [reply]
OpenType 的 TeX Gyre TermesX 其实很早就有了。但 `newtxtext` 在 v1.7 版本中才更新了在 XeLaTeX 和 LuaLaTeX 这类支持 Unicode 的引擎中通过 fontspec 使用 OpenType 版本的 TeX Gyre TermesX 字体。

实际上 `newtx` 包本身并不提供 `sf` 和 `tt` 字体，即使是在使用传统的 Type 1 字体模式下，`newtxtext` 设置的 `sf` 字体 `qhv` 实际就是 TeX Gyre Heros 的 Type 1 版本，而 `tt` 字体 `ntxtt` 来自于 `txfonts` 字体包。SJTUThesis 下个版本会转用 OpenType 版 TeX Gyre TermesX，并选择 OpenType 版的 TeX Gyre Heros 和  TeX Gyre Cursor 作为搭配。

**Comment by syvshc** [comment]
另外我在 tex.se 上获得了另一个设置的方法 https://tex.stackexchange.com/a/642789/180617 。我按照这个方法设置了一下字体，请问这样设置与 SJTUThesis 中的设置有什么优缺点呢？因为我最近也在为一些模板内的 `newtx` 系列宏包打补丁，想更多了解一下各处理方式的优劣。

```latex
\documentclass{article}
\usepackage[no-math]{fontspec}
\usepackage{newtxmath}
\setmainfont{TeXGyreTermesX}[
  Extension = .otf,
  UprightFont = *-Regular,
  BoldFont = *-Bold,
  ItalicFont = *-Italic,
  BoldItalicFont = *-BoldItalic,
  NFSSFamily = texgyretermesx,
]


\DeclareSymbolFont{operators}{TU}{texgyretermesx}{m}{n}
\DeclareMathAlphabet{\mathbf}{TU}{texgyretermesx}{b}{n}
\DeclareMathAlphabet{\mathit}{TU}{texgyretermesx}{m}{it}

\begin{document}
  English, \textbf{English}, \textit{English} $ \mathrm{English}, \mathbf{English}, \mathit{English}, \sin x $ 
\end{document}
```

![image](https://user-images.githubusercontent.com/38098591/166859367-65199929-1693-4e7f-9d9f-89cc305886f0.png)

**Reply by syvshc** [reply]
看起来模板内的处理方式更好一些，使得在数学模式下完全使用 type1 字体，可能兼容性会更好？

**Reply by AlexaraWu** [reply]
上面的代码是通过使用 NFSS 别名，重定义数学环境中的 `operators`（等同于 `\mathrm`）、`\mathit` 、`\mathbf` 使用 `fontspec` 设置 `TeXGyreTermesX`。我觉得没有什么优劣之分。SJTUThesis 中的设置更多是功能上的考虑。即我们实现了两个选项 `text-font` 和 `math-font`，使得模板内正文和数学环境中的字体设置互相独立。

**Reply by syvshc** [reply]
> 等同于 \mathrm 

我试了一下, `\DeclareMathAlphabet{\mathrm}{TU}{texgyretermesx}{m}{n}` 并不能重定义 `operators` 的字体, 但是` texdoc fntguide`  上对 `operator` 的定义是

![image](https://user-images.githubusercontent.com/38098591/166861827-0d05db25-9fd5-4571-8e0b-5f4ebebab459.png)

```latex
\documentclass{article}
\usepackage[no-math]{fontspec}
% \usepackage{newtxmath}
\setmainfont{TeXGyreTermesX}[
  Extension = .otf,
  UprightFont = *-Regular,
  BoldFont = *-Bold,
  ItalicFont = *-Italic,
  BoldItalicFont = *-BoldItalic,
  NFSSFamily = texgyretermesx,
]


% \DeclareSymbolFont{operators}{TU}{texgyretermesx}{m}{n}
\DeclareMathAlphabet{\mathrm}{TU}{texgyretermesx}{m}{n}
\DeclareMathAlphabet{\mathbf}{TU}{texgyretermesx}{b}{n}
\DeclareMathAlphabet{\mathit}{TU}{texgyretermesx}{m}{it}

\begin{document}
  English, \textbf{English}, \textit{English} $ \mathrm{English}, \mathbf{English}, \mathit{English}, \sin x $ 
\end{document}
```

![image](https://user-images.githubusercontent.com/38098591/166861883-b6f49e26-9774-4a46-8324-c7e2869cba46.png)

**Reply by syvshc** [reply]
嗯……我应该是找到了答案 : https://tex.stackexchange.com/a/219427/180617 , 我的问题解决了, 感谢您的回答. :)
